saw-core/src/SAWCore/Name.hs

```diff
@@ -285,6 +296,62 @@ scFreshNameURI nm i = fromMaybe (panic "scFreshNameURI" ["Failed to construct na
        }
 
 
+-- Display Name Environments --------------------------------------------------------
+
+-- | A 'DisplayNameEnv' encodes the mappings between display names (type
+-- 'Text') and internal names (type 'VarIndex'). Multiple display
+-- names may be associated with each internal name; these are stored
+-- in priority order, with preferred display names first.
+data DisplayNameEnv =
+  DisplayNameEnv
+  { displayNames :: !(IntMap [Text]) -- Keyed by VarIndex; preferred names come first.
+  , displayIndexes :: !(Map Text IntSet)
+  }
+-- Invariants: The 'displayNames' and 'displayIndexes' maps should be
+-- inverses of each other. That is, 'displayNames' maps @i@ to a list
+-- containing @x@ if and only if 'displayIndexes' maps @x@ to a set
+-- containing @i@.
+
+emptyDisplayNameEnv :: DisplayNameEnv
+emptyDisplayNameEnv = DisplayNameEnv mempty mempty
+
+-- | Extend a 'DisplayNameEnv' by providing new 'Text' display names
+-- for a given 'VarIndex'. Display names should be provided in
+-- priority order, with preferred names first. If the 'VarIndex'
+-- already exists in the map, then add the new display names with
+-- higher priority while keeping the old ones.
+extendDisplayNameEnv :: VarIndex -> [Text] -> DisplayNameEnv -> DisplayNameEnv
+extendDisplayNameEnv i aliases env =
+  DisplayNameEnv
+  { displayNames = IntMap.insertWith List.union i aliases (displayNames env)
+  , displayIndexes = foldr insertAlias (displayIndexes env) aliases
+  }
+  where
+    insertAlias :: Text -> Map Text IntSet -> Map Text IntSet
+    insertAlias x m = Map.insertWith IntSet.union x (IntSet.singleton i) m
+
+resolveDisplayName :: DisplayNameEnv -> Text -> [VarIndex]
+resolveDisplayName env x =
+  case Map.lookup x (displayIndexes env) of
+    Nothing -> []
+    Just s -> IntSet.elems s
+
+-- | Return the first display name for the given 'VarIndex' that is
+-- unambiguous in the naming environment. If there is no unambiguous
+-- alias, then return 'Nothing'.
+bestDisplayName :: DisplayNameEnv -> VarIndex -> Maybe Text
+bestDisplayName env i =
+  do aliases <- IntMap.lookup i (displayNames env)
+     go aliases
+  where
+    go [] = Nothing
+    go (x : xs) =
+      case Map.lookup x (displayIndexes env) of
+        Nothing -> go xs -- should never happen
```

Comment:
If this shouldn't happen, then should this case panic instead?

Reply:
Yes, it probably should. If `env` satisfies the documented invariant, the lookup will always succeed. I'll change that to a panic.

Reply:
Done in 4655abf.